libsolidity/parsing/Parser.cpp

```diff
@@ -1858,6 +1842,41 @@ ASTPointer<TypeDefinition> Parser::parseTypeDefinition()
 	);
 }
 
+ASTPointer<Builtin> Parser::parseBuiltin()
+{
+	solAssert(m_experimentalSolidityEnabledInCurrentSourceUnit);
+	ASTNodeFactory nodeFactory(*this);
+	expectToken(Token::Builtin);
+	expectToken(Token::LParen);
+
+	auto builtin = nodeFactory.createNode<Builtin>(
+		std::make_shared<std::string>(m_scanner->currentLiteral()),
+		m_scanner->currentLocation()
+	);
+
+	expectToken(Token::StringLiteral);
+
+	if(m_scanner->currentToken() == Token::Comma)
+	{
+		expectToken(Token::Comma);
+		// TODO: Consider if other ASTNode would be more adequate (MemberAccess?)
```

Comment:
I'd actually say it should plainly be an ``Expression`` - there is actually no need to restrict this to a type class member function - it can just as well be *any* kind of function whatsoever (so a free function or - once we had them - a lambda function)...

We actually don't even need to resolve the expression in any way - we can just store a map from operator to the defining expression and then just jump to that expression during type inference directly (that way, *any* expression that can be called should work just fine)

Reply:
So in particular: if you make this an expression, you can just change ``TypeInference::visit(BinaryOperation const& _binaryOperation)`` (the ``ExpressionContext::Term`` case), to fetch the `Expression` associated to ``_binaryOperation.getOperator()``, then have ``TypeInference`` *visit* that expression (if it doesn't have a ``.type`` annotation yet) and then use its ``.type`` annotation as ``functionType`` directly (instead of fetching a class member function)